File: android/app/src/main/java/com/agenticandroid/Protocol.kt
package com.agenticandroid

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

/**
 * Wire protocol — the Kotlin side of THE contract. Byte-for-byte compatible with backbone/src/protocol.ts.
 * Outer Envelope is relay-visible; Inner (inside `enc`) is E2E-encrypted and never seen by the relay.
 */
const val PROTOCOL_VERSION = 1

@Serializable
data class Envelope(
    val v: Int = PROTOCOL_VERSION,
    val id: String,
    val from: String,
    val to: String,
    val ts: Long,
    val enc: String,
)

@Serializable
data class TypedError(val code: String, val message: String, val retriable: Boolean = false)

/** The four inner message kinds. Discriminator field is "type" to match the TS InnerMessage union. */
@Serializable
sealed class Inner {
    @Serializable @SerialName("request")
    data class Request(val method: String, val params: JsonObject = JsonObject(emptyMap())) : Inner()

    @Serializable @SerialName("response")
    data class Response(
        @SerialName("reply_to") val replyTo: String,
        val status: String, // "ok" | "error"
        val result: JsonElement? = null,
        val error: TypedError? = null,
    ) : Inner()

    @Serializable @SerialName("event")
    data class Event(val topic: String, val data: JsonObject = JsonObject(emptyMap())) : Inner()

    @Serializable @SerialName("ack")
    data class Ack(val ack: String) : Inner()
}

/** Note: TS uses "reply_to"; @SerialName above keeps the wire key identical. */
val ProtocolJson = Json {
    classDiscriminator = "type"
    ignoreUnknownKeys = true
    encodeDefaults = true
}

fun encodeInner(inner: Inner): String = ProtocolJson.encodeToString(Inner.serializer(), inner)
fun parseInner(json: String): Inner = ProtocolJson.decodeFromString(Inner.serializer(), json)
fun encodeEnvelope(env: Envelope): String = ProtocolJson.encodeToString(Envelope.serializer(), env)
fun parseEnvelope(json: String): Envelope = ProtocolJson.decodeFromString(Envelope.serializer(), json)

File: backbone/src/protocol.ts
/**
 * Wire protocol — THE contract. Everything else is an adapter behind this.
 *
 * Two layers:
 *  - Envelope (cleartext): what the relay sees. Just enough to route: from/to/id/ts + opaque `enc`.
 *  - Inner message (inside `enc`, E2E-encrypted): the 4 message kinds. The relay never sees these.
 *
 * Relay-control frames (hello/challenge/auth/welcome/error) are NOT envelopes — they ride the same
 * socket during the handshake and are told apart by the `ctl` field.
 */
import { z } from "zod";

export const PROTOCOL_VERSION = 1;

// ---------- Outer envelope (relay-visible) ----------
export const Envelope = z.object({
  v: z.literal(PROTOCOL_VERSION),
  id: z.string().min(1), // unique per message; responses correlate via inner.reply_to === this id
  from: z.string().min(1), // sender fingerprint
  to: z.string().min(1), // destination fingerprint
  ts: z.number().int(),
  enc: z.string().min(1), // base64 ciphertext of an InnerMessage
});
export type Envelope = z.infer<typeof Envelope>;

// ---------- Inner messages (E2E-encrypted; 4 kinds) ----------
export const TypedError = z.object({
  code: z.string(), // e.g. CAMERA_IN_USE, PERMISSION_NOT_GRANTED, CONSENT_DENIED, UNKNOWN_METHOD
  message: z.string(),
  retriable: z.boolean().default(false),
});
export type TypedError = z.infer<typeof TypedError>;

export const RequestMsg = z.object({
  type: z.literal("request"),
  method: z.string(), // e.g. "phone.ring", "camera.capture", "schedule"
  params: z.record(z.unknown()).default({}),
});

export const ResponseMsg = z.object({
  type: z.literal("response"),
  reply_to: z.string(), // envelope id of the request being answered
  status: z.enum(["ok", "error"]),
  result: z.unknown().optional(),
  error: TypedError.optional(),
});

export const EventMsg = z.object({
  type: z.literal("event"),
  topic: z.string(), // e.g. "user_message", "wakeword", "notification.posted", "task.result"
  data: z.record(z.unknown()).default({}),
});

export const AckMsg = z.object({
  type: z.literal("ack"),
  ack: z.string(), // envelope id being acknowledged
});

export const InnerMessage = z.discriminatedUnion("type", [
  RequestMsg,
  ResponseMsg,
  EventMsg,
  AckMsg,
]);
export type InnerMessage = z.infer<typeof InnerMessage>;
export type RequestMsg = z.infer<typeof RequestMsg>;
export type ResponseMsg = z.infer<typeof ResponseMsg>;
export type EventMsg = z.infer<typeof EventMsg>;
export type AckMsg = z.infer<typeof AckMsg>;

// ---------- Relay control frames (handshake) ----------
export const ControlFrame = z.discriminatedUnion("ctl", [
  z.object({ ctl: z.literal("hello"), fp: z.string(), edpub: z.string() }),
  z.object({ ctl: z.literal("challenge"), nonce: z.string() }),
  z.object({ ctl: z.literal("auth"), sig: z.string() }),
  z.object({ ctl: z.literal("welcome") }),
  z.object({ ctl: z.literal("error"), message: z.string() }),
]);
export type ControlFrame = z.infer<typeof ControlFrame>;

// ---------- Parsing helpers (trust boundary: validate everything off the wire) ----------

/** Parse a raw WS text frame into either a control frame or an envelope. Throws on garbage. */
export function parseFrame(raw: string): { kind: "ctl"; frame: ControlFrame } | { kind: "env"; env: Envelope } {
  const obj = JSON.parse(raw);
  if (obj && typeof obj === "object" && "ctl" in obj) {
    return { kind: "ctl", frame: ControlFrame.parse(obj) };
  }
  return { kind: "env", env: Envelope.parse(obj) };
}

export function parseInner(json: string): InnerMessage {
  return InnerMessage.parse(JSON.parse(json));
}

export function encodeInner(msg: InnerMessage): string {
  return JSON.stringify(InnerMessage.parse(msg));
}

let _seq = 0;
/** Monotonic-ish unique message id. Not security-sensitive (routing only). */
export function newId(prefix = "m"): string {
  _seq = (_seq + 1) & 0xffffff;
  return `${prefix}_${Date.now().toString(36)}_${_seq.toString(36)}`;
}
